### PR TITLE
NAS visualization

### PR DESCRIPTION
**Pull request by ultmaster**
~This PR is open for code comparison against master. Not ready for review yet.~

Working items:

- [x] Makefile
- [x] Fix UI fail bug when key is missing.

Might do if I have time:

- [x] Use proto graph and log in protobuf (instead of NodePy)
- [x] Configurable node elimination on UI
- [x] Configurable node collapse on UI
- [x] Display attributes of ops on UI

I will post some preview here when I think it's ready.

**Ready for review**

Deferred items:

- [ ] Add a summary of candidates along with weights in expansion panel.
- [ ] Layer choice candidate infer is still buggy. Need investigation.
- [ ] Add panzoom.
- [ ] Add focus mode showing mutable nodes only.
- [ ] Support for build and entrypoint in non-source build. (need help)
- [ ] Hide primitive nodes (like ListConstruct)
- [ ] Merge multiple edges between two nodes/clusters.
- [ ] Design might need improving (theme, graph style). *I know it's ugly, but I can't bend the curves. I tried...*
- [ ] Responsive width & height.
- [x] Clean up unused dependencies.
- [ ] Hide nodes with weight 0.

*These items might be done in separate PRs.*

image

image

image

image

How to test:

Change DARTS search into something like this:

```python
class MyTrainer(DartsTrainer):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.status_writer = open("log", "w")

    def _logits_and_loss(self, X, y):
        self.mutator.reset()
        logits = self.model(X)
        loss = self.loss(logits, y)
        print(json.dumps(self.mutator.status()), file=self.status_writer)
        self.status_writer.flush()
        return logits, loss


model = CNN(32, 3, 16, 10, 8)
model.cuda()
mutator = DartsMutator(model)
vis_graph = mutator.graph(torch.randn((1, 3, 32, 32)).cuda())
with open("graph.json", "w") as f:
    json.dump(vis_graph, f)

dataset_train, dataset_valid = datasets.get_dataset("cifar10")
criterion = nn.CrossEntropyLoss()
optim = torch.optim.SGD(model.parameters(), 0.025, momentum=0.9, weight_decay=3.0E-4)
trainer = MyTrainer(model=model,
                    mutator=mutator,
                    loss=criterion,
                    metrics=lambda output, target: accuracy(output, target, topk=(1,)),
                    optimizer=optim,
                    num_epochs=1,
                    dataset_train=dataset_train,
                    dataset_valid=dataset_valid,
                    batch_size=64,
                    log_frequency=10,
                    arc_learning_rate=0.1,
                    unrolled=False)
trainer.train()
```

This will write the graph to `graph.json` before starts, and write each step as a line into a file called `log`.

Launch `nnictl webui nas --logdir /path/to/the/directory/containing/two/files`. Find your UI at port ~6667~ 6060.

### CODE REVIEW

File: src/nasui/src/Chart.tsx [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    selector: 'node',
+    style: {
+      'shape': 'round-rectangle',
+      'width': 'label',
+      'height': 'label',
+      'content': 'data(label)',
+      'text-wrap': 'wrap',
+      'text-valign': 'center',
+      'text-halign': 'center',
+      'font-family': '"Roboto", "Helvetica", "Arial", sans-serif',
```
[the remark this answers is not in the document]
Comment (ultmaster) [reply src/nasui/src/Chart.tsx]: I'm aligning with tensorboard. So I used Material-UI + Roboto.

File: src/nasui/src/Chart.tsx [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+  },
+  {
+    selector: 'node.compound',
+    style: {
+      'shape': 'roundrectangle',
+      'text-valign': 'top',
+      'padding-top': '30px',
+      'padding-bottom': '10px',
+      'padding-right': '10px',
+      'padding-left': '10px',
```
[the remark this answers is not in the document]
Comment (ultmaster) [reply src/nasui/src/Chart.tsx]: I don't think so. Cytoscape use its own style format. I tried to merge them and it didn't work.

File: src/nasui/src/Chart.tsx [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+      'content': 'data(label)',
+      'text-wrap': 'wrap',
+      'text-valign': 'center',
+      'text-halign': 'center',
+      'font-family': '"Roboto", "Helvetica", "Arial", sans-serif',
+      'font-size': 20,
+      'padding-left': '8px',
+      'padding-right': '8px',
+      'padding-top': '8px',
+      'padding-bottom': '8px',
```
[the remark this answers is not in the document]
Comment (ultmaster) [reply src/nasui/src/Chart.tsx]: Same

File: src/nasui/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+    const handleSettingsChange = (name: string) => (event: React.ChangeEvent<HTMLInputElement>) => {
+      this.setState({
+        ...this.state,
+        [name]: event.target.checked
+      }, () => {
+        this.setState({
+          graph: new Graph(this.state.graphData, this.state.hideSidechainNodes),
+        })
+      });
+    };
```
[the remark this answers is not in the document]
Comment (ultmaster) [reply src/nasui/src/App.tsx]: Good suggestion. Will do.

File: src/nasui/package.json [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    "@testing-library/user-event": "^7.1.2",
+    "@types/cytoscape": "^3.14.0",
+    "@types/d3-graphviz": "^2.6.3",
+    "@types/dagre-d3": "^0.4.39",
+    "@types/graphlib-dot": "^0.6.1",
+    "@types/jest": "^24.0.0",
+    "@types/lodash": "^4.14.149",
+    "@types/node": "^12.0.0",
+    "@types/react": "^16.9.0",
+    "@types/react-dom": "^16.9.0",
```
[the remark this answers is not in the document]
Comment (ultmaster) [reply src/nasui/package.json]: It's in future item: `Clean up unused dependencies.`

File: src/nasui/src/Chart.tsx [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+    });
+    return graphElements;
+  }
+
+  private applyMutableWeight() {
+    const { graph, activation } = this.props;
+    if (graph === undefined || activation === undefined)
+      return;
+    const weights = graph.weightFromMutables(activation);
+    console.log(weights.size);
```
[the remark this answers is not in the document]
Comment (ultmaster) [reply src/nasui/src/Chart.tsx]: I will delete all `console.log` before merge.

File: src/nasui/src/App.tsx [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+          snackbarOpen: graph.nodes.length > 100
+        });
+      });
+  }
+
+  private renderExpansionPanel() {
+    const { classes } = this.props;
+    const { selectedNode, graph } = this.state;
+    if (graph === undefined)
+      return null;
```
Comment (SparkSnail) [inline src/nasui/src/App.tsx]: what is the difference between `null` and `undefined`?

File: src/nasui/src/graphUtils.ts [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+      })
+    })
+    this.build();
+
+    if (eliminateSidechains) {
+      this.eliminateSidechains();
+    }
+    this.defaultExpandSet = this.getDefaultExpandSet(graphData.mutable);
+    this.mutableEdges = this.inferMutableEdges(graphData.mutable);
+    console.log(this);
```
Comment (SparkSnail) [inline src/nasui/src/graphUtils.ts]: remove console.log()

File: src/nasui/src/react-app-env.d.ts [hunk 9]
```diff
@@ -0,0 +1 @@
+/// <reference types="react-scripts" />
```
Comment (SparkSnail) [inline src/nasui/src/react-app-env.d.ts]: Is this file useful?

File: src/nasui/src/serviceWorker.ts [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+              if (config && config.onSuccess) {
+                config.onSuccess(registration);
+              }
+            }
+          }
+        };
+      };
+    })
+    .catch(error => {
+      console.error('Error during service worker registration:', error);
```
Comment (SparkSnail) [inline src/nasui/src/serviceWorker.ts]: use `Alert(error)`?

File: src/nasui/src/setupTests.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+// jest-dom adds custom jest matchers for asserting on DOM nodes.
```
Comment (SparkSnail) [inline src/nasui/src/setupTests.ts]: Is this useful? It seems that there is no main logic code.

File: tools/nni_cmd/nnictl.py [hunk 12]
```diff
@@ -158,6 +158,10 @@ def parse_args():
     parser_webui_url = parser_webui_subparsers.add_parser('url', help='show the url of web ui')
     parser_webui_url.add_argument('id', nargs='?', help='the id of experiment')
     parser_webui_url.set_defaults(func=webui_url)
+    parser_webui_nas = parser_webui_subparsers.add_parser('nas', help='show nas ui')
+    parser_webui_nas.add_argument('--port', default=6060, type=int)
```
Comment (SparkSnail) [inline tools/nni_cmd/nnictl.py]: add `help=` to explain the argement.

File: src/nasui/src/serviceWorker.ts (hunk 10 above)
Comment (ultmaster) [reply src/nasui/src/serviceWorker.ts]: This file is auto-generated. I don't want to modify it.

File: src/nasui/src/react-app-env.d.ts (hunk 9 above)
Comment (ultmaster) [reply src/nasui/src/react-app-env.d.ts]: Yes.

File: src/nasui/src/setupTests.ts (hunk 11 above)
Comment (ultmaster) [reply src/nasui/src/setupTests.ts]: It's auto-generated. Removing it for now.

File: tools/nni_cmd/nnictl.py (hunk 12 above)
Comment (ultmaster) [reply tools/nni_cmd/nnictl.py]: added

File: src/nasui/src/App.tsx (hunk 7 above)
Comment (ultmaster) [reply src/nasui/src/App.tsx]: React suggests `null` when returning a None element. Fixed. Thanks.

File: src/nasui/src/graphUtils.ts (hunk 8 above)
Comment (ultmaster) [reply src/nasui/src/graphUtils.ts]: done

File: src/nasui/src/react-app-env.d.ts (hunk 9 above)
Comment (QuanluZhang) [reply src/nasui/src/react-app-env.d.ts]: @ultmaster better to explain why

Comment (ultmaster) [reply src/nasui/src/react-app-env.d.ts]: It's explained in https://github.com/facebook/create-react-app/issues/6560. I guess we should respect the practice of create-react-app.

File: src/sdk/pynni/nni/nas/pytorch/mutator.py [hunk 13]
```diff
@@ -57,6 +60,70 @@ def export(self):
...
+                {"type": self.model.__class__.__name__, "name": ""}
+            ]
+            m = self.model
+            for module in modules:
+                m = getattr(m, module)
+                path.append({
+                    "type": m.__class__.__name__,
+                    "name": module
+                })
+            result["mutable"][mutable.key].append(path)
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/mutator.py]: better to add comments to explain this part

File: tools/nni_cmd/nnictl.py [hunk 14]
```diff
@@ -203,7 +207,7 @@ def parse_args():
     parser_tensorboard_start = parser_tensorboard_subparsers.add_parser('start', help='start tensorboard')
     parser_tensorboard_start.add_argument('id', nargs='?', help='the id of experiment')
     parser_tensorboard_start.add_argument('--trial_id', '-T', dest='trial_id', help='the id of trial')
-    parser_tensorboard_start.add_argument('--port', dest='port', default=6006, help='the port to start tensorboard')
+    parser_tensorboard_start.add_argument('--port', dest='port', default=6060, help='the port to start tensorboard')
```
Comment (QuanluZhang) [inline tools/nni_cmd/nnictl.py]: why change the default value?

**Comment by QuanluZhang** [comment]
@ultmaster could you write doc for it, including how to change user code, how to start the nas ui.

**Comment by ultmaster** [comment]
> @ultmaster could you write doc for it, including how to change user code, how to start the nas ui.

I introduced briefly how to use it in the description of this PR, but it's for test only.

The docs will be in next release. This is because currently, the only way to launch it is to **build from source and launch it with nnictl under nni directly**, which is because, release integration is not done.

BTW, as discussed in the last meeting, I need help to integrate it into release (including changing all the Makefiles and make it work on pypi version).

File: src/sdk/pynni/nni/nas/pytorch/mutator.py (hunk 13 above)
Comment (ultmaster) [reply src/sdk/pynni/nni/nas/pytorch/mutator.py]: Sure

File: tools/nni_cmd/nnictl.py (hunk 14 above)
Comment (ultmaster) [reply tools/nni_cmd/nnictl.py]: It's an accident. Sorry.

**Pull request merged**